src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ApplicationPropertyPage/ApplicationIconValueProvider.cs

```diff
@@ -0,0 +1,59 @@
+
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.IO;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Properties.Package
+{
+    // Sets the ApplicationIcon property to the path to the .ico file. This path is relative
+    // to the project file directory. If the selected file is not within the project directory
+    // tree, the file will be copied into the project file directory. Previously copied files
+    // are not deleted when a new file is selected. ApplicationIcon allows for both full paths
+    // and relative paths anywhere on disk, but we are not utilizing that in this implementation.
+    // The icon file does not need to be included in the project (such as a None/Content item).
+    [ExportInterceptingPropertyValueProvider(ApplicationIconPropertyName, ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+    internal sealed class ApplicationIconValueProvider : InterceptingPropertyValueProviderBase
+    {
+        private const string ApplicationIconPropertyName = "ApplicationIcon";
+
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IFileSystem _fileSystem;
+
+        [ImportingConstructor]
+        public ApplicationIconValueProvider(UnconfiguredProject unconfiguredProject, IFileSystem fileSystem)
+        {
+            _unconfiguredProject = unconfiguredProject;
+            _fileSystem = fileSystem;
+        }
+
+        public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+        {
+            // Allows the user to clear the application icon.
+            if (string.IsNullOrEmpty(unevaluatedPropertyValue))
+            {
+                return null;
+            }
+
+            // If selected path is invalid, ignore and return previous value.
+            if (!File.Exists(unevaluatedPropertyValue))
+            {
+                return await defaultProperties.GetUnevaluatedPropertyValueAsync(propertyName);
+            }
+
+            // Copy the file into the project directory.
+            // This isn't a limitation of ApplicationIcon; it is to simply mimic the legacy property page interaction.
+            if (_unconfiguredProject.IsOutsideProjectDirectory(unevaluatedPropertyValue))
+            {
+                string fileName = Path.GetFileName(unevaluatedPropertyValue);
+                string destination = Path.Combine(Path.GetDirectoryName(_unconfiguredProject.FullPath), fileName);
+                _fileSystem.CopyFile(unevaluatedPropertyValue, destination, overwrite: true);
+                return fileName;
```

Comment:
Shouldn't this return `destination`? I don't understand why we copy the file to the project directory if we don't then use that path.

Reply:
The path to the file (what is put into `ApplicationIcon`) is relative to the project directory. Therefore, the filename (example.ico) is the path to the file, which is why it is returned. This is how the legacy property pages worked too. `destination` is simply needed since this `IFileSystem` API needs 2 explicit paths to do the copy action. I tried just putting the directory as the destination at first, and it doesn't have the smarts to recognize that I wanted to copy a file into a folder.